### PR TITLE
[WIP] Finder with base message

### PR DESCRIPTION
**Pull request by danielmarbach**
Based on top of #3996

### CODE REVIEW

**Comment by danielmarbach** [comment]
@SimonCropp I just realized that I might be wrong with the finders. Look for example at the following:

```
        class SagaWithCustomFinderForMessageItDoesntHandle : Saga<SagaWithCustomFinderForMessageItDoesntHandle.SagaData>,
            IAmStartedByMessages<SomeMessage>
        {
            protected override void ConfigureHowToFindSaga(SagaPropertyMapper<SagaData> mapper)
            {
                mapper.ConfigureMapping<SomeMessage>(msg => msg.SomeProperty).ToSaga(saga => saga.SomeProperty);
            }

            public Task Handle(SomeMessage message, IMessageHandlerContext context)
            {
                throw new NotImplementedException();
            }

            public class Finder : IFindSagas<SagaData>.Using<OtherMessage>
            {
                public Task<SagaData> FindBy(OtherMessage message, SynchronizedStorageSession storageSession, ReadOnlyContextBag context)
                {
                    return Task.FromResult(default(SagaData));
                }
            }

            public class SagaData : ContainSagaData
            {
                public int SomeProperty { get; set; }
            }

            public class OtherMessage : IMessage
            {
                public int SomeProperty { get; set; }
            }
        }
```

This will now throw with

> A custom IFindSagas and an existing mapping where found for message '{messageType.FullName}'. Either remove the message mapping for remove the finder. Finder name '{finderType.FullName}'.

since both message messages inherit from `IMessage`.

**Comment by timbussmann** [comment]
@danielmarbach is this still in progress?

**Comment by danielmarbach** [comment]
This seems to be a pandora box. Not willing to open it before RTM. What's your thoughts @Particular/nservicebus-maintainers ?

**Comment by andreasohlund** [comment]
Needs to be retargeted against develop

**Comment by andreasohlund** [comment]
Strike that, I re-targeted it

**Comment by danielmarbach** [comment]
Not sure what we want to do with this to be honest

**Comment by danielmarbach** [comment]
@SimonCropp do you have any ideas what we should do here?

https://github.com/Particular/NServiceBus/pull/4001#issuecomment-236850251

and https://github.com/Particular/NServiceBus/pull/3996#issuecomment-236478779

I'm totally unsure what the right behavior should be here. Maybe @DavidBoike can help out?

**Comment by DavidBoike** [comment]
IMO, the important thing is that the messages handled by the saga have a valid mapping. If that comes from a base class to that message, that shouldn't matter, right? So then just checking the metadata with an IsAssignableTo rather than Type equals would seem to work. That's not going to make the saga suddenly handle messages it otherwise wouldn't have.

**Comment by danielmarbach** [comment]
@davidboike the mapping check then crashes with

A custom IFindSagas and an existing mapping where found for message '{messageType.FullName}'. Either remove the message mapping for remove the finder. Finder name '{finderType.FullName}'.

File: src/NServiceBus.Core/Sagas/SagaMetadata.cs [hunk 1]
```diff
@@ -258,7 +255,7 @@ static void ApplyScannedFinders(SagaMapper mapper, Type sagaEntityType, IEnumera
                         throw new Exception(error);
                     }
 
-                    var existingMapping = mapper.Mappings.SingleOrDefault(m => m.MessageType == messageType);
+                    var existingMapping = mapper.Mappings.SingleOrDefault(m => m.MessageType.IsAssignableFrom(m.MessageType));
                     if (existingMapping != null)
                     {
                         var bothMappingAndFinder = $"A custom IFindSagas and an existing mapping where found for message '{messageType.FullName}'. Either remove the message mapping for remove the finder. Finder name '{finderType.FullName}'.";
```
Comment (timbussmann) [inline src/NServiceBus.Core/Sagas/SagaMetadata.cs]: this exception mapping requires some improvements. where -> were. for -> or.

**Comment by timbussmann** [comment]
@danielmarbach reading through the comments and the original issue, I'm not really able to follow the problems. There seems to be some conflicts with custom saga mappers, can you provide more details in what scenarios this causes an exception now? Is this also a problem when not using a custom finder (as I haven't read anything about one in @DavidBoike's comment https://github.com/Particular/NServiceBus/pull/4001#issuecomment-254240179.

**Pull request closed without merging**